crates/datasources/src/common/sink/parquet.rs

```diff
@@ -14,20 +17,21 @@ use datafusion::physical_plan::{DisplayAs, DisplayFormatType, SendableRecordBatc
 use futures::StreamExt;
 use object_store::path::Path as ObjectPath;
 use object_store::ObjectStore;
+use tokio::io::AsyncWrite;
+use tokio::task::JoinSet;
+
+use crate::common::sink::write::parquet::{output_single_parquet_file_parallelized, ParallelParquetWriterOptions};
+use crate::common::sink::write::{AbortableWrite, MultiPart};
+
+use super::write::demux::start_demuxer_task;
 
 const BUFFER_SIZE: usize = 8 * 1024 * 1024;
 
 #[derive(Debug, Clone)]
 pub struct ParquetSinkOpts {
```

Comment:
my inclination was that we'd be able to do this _above_ the level of the individual formats. so that you could have hive partitioning for any supported output (file?) format.

Reply:
Gotcha.

I'll have a think about what that might look like. I suspect with a generic approach there would be some performance trade offs for parquet but let me get back to you.

Reply:
Interesting. I would imagine that generic approaches to reading wouldn't work super well with _reading_ parquet but I'd imagine that writing data might be roughly the same. What were you thinking the impact would be?

Reply:
Yes I think you're right. There's some code in the DF repo that parallelizes single-file  parquet encoding that I thought may interact with the hive stuff but it looks like it's independent - which does make sense now I think about it.

I've updated the PR to introduce an abstraction that accepts a data sink "producer" that will be called on each new partitioned file with the partition path. So each format just needs to implement the SinkProducer trait and this will unlock hive partitioning. Please see `hive_partitioning.rs`.

At the moment I've just added Parquet (see copy_to.rs) but it should be easy to add for other formats. Please let me know what you think of this approach. 

I've also removed the single-file parallelization code for parquet that I'd brought over from DF as it's not needed for hive partitioning (though may be worth looking at adding that later as an optimization).